pkg/instancegroups/rollingupdate.go

```diff
@@ -17,177 +17,319 @@ limitations under the License.
+	}
+	// @check if the batch size is at least one
+	if strategy.Batch <= 0 {
+		strategy.Batch = 1
+	}
+	// @check if the batch size has been overrided by options
+	if c.Batch > 0 {
+		strategy.Batch = c.Batch
+	}
+	// @check if cloudonly and adject the drain options
+	if c.CloudOnly {
+		strategy.Drain = false
+	}
+	if c.Drain != nil {
+		strategy.Drain = *c.Drain
 	}
 
-	// Upgrade nodes, with greater parallelism
-	{
-		var wg sync.WaitGroup
+	// @step: work out the interval for this role
+	if strategy.Interval == nil {
+		switch role {
+		case api.InstanceGroupRoleBastion:
+			strategy.Interval = &metav1.Duration{Duration: c.BastionInterval}
+		case api.InstanceGroupRoleMaster:
+			strategy.Interval = &metav1.Duration{Duration: c.MasterInterval}
+		case api.InstanceGroupRoleNode:
+			strategy.Interval = &metav1.Duration{Duration: c.NodeInterval}
+		}
+	}
 
-		// We run nodes in series, even if they are in separate instance groups
-		// typically they will not being separate instance groups. If you roll the nodes in parallel
-		// you can get into a scenario where you can evict multiple statefulset pods from the same
-		// statefulset at the same time. Further improvements needs to be made to protect from this as
-		// well.
+	// @step: check the stratergy is compatible with the role
+	switch role {
+	case api.InstanceGroupRoleMaster, api.InstanceGroupRoleBastion:
+		if strategy.Name != api.DefaultRollout {
+			return fmt.Errorf("rollout strategy: %s is not supported for role: %s", strategy.Name, role)
+		}
+	}
 
-		wg.Add(1)
+	// @check the rollout stratergy
+	switch strategy.Name {
+	case api.DuplicateRollout, api.ScaleUpRollout:
+		if role != api.InstanceGroupRoleNode {
+			return c.Errorf("rollout strategy: %s is only supported on node instancegroups", api.DuplicateRollout)
+		}
+		if len(group.NeedUpdate) > 0 {
+			c.Infof("toggling the force flag given we are using a %s strategy", strategy.Name)
+			c.Force = true
+		}
+	}
 
-		go func() {
-			for k := range nodeGroups {
-				resultsMutex.Lock()
-				results[k] = fmt.Errorf("function panic nodes")
-				resultsMutex.Unlock()
-			}
+	// @check try to ensure that if everything requires a update and the batch size is
+	// bigger than the group, we leave an instance active
+	if len(group.Ready) == 0 {
+		if strategy.Batch > 1 && strategy.Batch >= len(group.NeedUpdate) {
+			strategy.Batch = strategy.Batch - 1
+			c.Infof("adjusting batch size to: %d to keep availability Ready/Update (0/%d)",
+				strategy.Batch, len(group.NeedUpdate))
+		}
+	}
 
-			defer wg.Done()
+	return nil
+}
 
-			for k, group := range nodeGroups {
-				g, err := NewRollingUpdateInstanceGroup(c.Cloud, group)
-				if err == nil {
-					err = g.RollingUpdate(c, instanceGroups, false, c.NodeInterval, c.ValidationTimeout)
-				}
+// IsGroupUpdating checks to see if the instancegroup is going to rollout
+func (c *RollingUpdateCluster) IsGroupUpdating(name string) bool {
+	if len(c.InstanceGroups) <= 0 {
```

Comment:
You don't need the length check, I don't think (?).  (For example, you can loop over a nil slice)